### PR TITLE
Fix handling of invalid inputs (<= 0) to `GridEvenShare` and adjust handling of `num_items == 0` on the caller side

### LINKED ISSUE #5435: [BUG]: `GridEvenShare` does not work properly when `num_items == 0`.
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Runtime Error

### Component

CUB

### Describe the bug

Passing `num_items == 0` to `GridEvenShare::DispatchInit` used to lead to division by zero which is undefined behavior. In release mode (on my machine), this currently leads to `grid_size == 0`, which leads to an invalid configuration argument error. In debug mode we get a `SIGFPE - Floating point error signal`.

In the atomic reduce PR, I changed it so that `total_tiles` is at least 1 https://github.com/NVIDIA/cccl/pull/4961/commits/f429931b445ec6bc2c568471e55fb08268b32064. Since `GridEvenShare` is part of the public API, and this changes its behavior, we need to decide more carefully how we should handle `num_items == 0`.

### How to Reproduce

Undo the change to `GridEvenShare` in the commit mentioned above and build and run `catch2_test_device_reduce_nondeterministic`.

### Expected behavior

We need to add tests specifically for `GridEvenShare` and decide how to handle corner cases like `num_items == 0`.

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by NaderAlAwar**
## Description

closes #5435

This PR also adds `GridEvenShare` tests.

No performance difference on H200.

## Checklist
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/test/catch2_test_grid_even_share.cu [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#include <cub/grid/grid_mapping.cuh>
+
+#include <cuda/std/__algorithm/min.h>
+#include <cuda/std/type_traits>
+
+#include <c2h/catch2_test_helper.h>
+#include <c2h/generators.h>
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_grid_even_share.cu]: Important: this is not needed, since we are not testing any kernel launches
```suggestion
```

File: cub/test/catch2_test_grid_even_share.cu [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }();
+
+  const int max_grid_size = GENERATE_COPY(values({-1, 0, 1}));
+  const int tile_items    = GENERATE_COPY(values({-1, 0, 1}));
+
+  // Skip if all parameters are positive (covered by the normal operation test)
+  if (num_items > 0 && max_grid_size > 0 && tile_items > 0)
+  {
+    return;
+  }
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_grid_even_share.cu:39]: Q: Does it harm to run this test? If not, I would just let it run. It does not look like this will consume a lot more resources.

File: cub/test/catch2_test_grid_even_share.cu [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  grid_share.DispatchInit(num_items, max_grid_size, tile_items);
+
+  REQUIRE(grid_share.num_items == 0);
+  REQUIRE(grid_share.grid_size == 0);
+  REQUIRE(grid_share.block_offset == 0);
+  REQUIRE(grid_share.block_end == 0);
```
Comment (bernhardmgruber) [inline cub/test/catch2_test_grid_even_share.cu]: Suggestion: I would use `CHECK` here because in case of a failure, it may be easier to have all the failed checks visible instead of just the first one:
```suggestion
  CHECK(grid_share.num_items == 0);
  CHECK(grid_share.grid_size == 0);
  CHECK(grid_share.block_offset == 0);
  CHECK(grid_share.block_end == 0);
```
Applies to the next test below as well.

File: cub/test/catch2_test_grid_even_share.cu (hunk 1 above)
Comment (NaderAlAwar) [reply cub/test/catch2_test_grid_even_share.cu]: My rationale here is that because `GridEvenShare::DispatchInit` is called in the dispatch layer, the code can be called in different launch contexts. I can remove if you think this is necessary

File: cub/test/catch2_test_grid_even_share.cu (hunk 2 above)
Comment (NaderAlAwar) [reply cub/test/catch2_test_grid_even_share.cu:39]: I want to leave this test specifically for invalid inputs. Allowing this test with valid inputs means that I would have to change the assertions, while the way it is currently explicitly shows that passing invalid inputs results in zeroing its members.

File: cub/test/catch2_test_grid_even_share.cu (hunk 1 above)
Comment (bernhardmgruber) [reply cub/test/catch2_test_grid_even_share.cu]: But you are not declaring a launch helper with `DECLARE_LAUNCH_WRAPPER`, which regards this macro and launches a kernel in 3 different ways. I see no point in compiling this test trice.

Comment (NaderAlAwar) [reply cub/test/catch2_test_grid_even_share.cu]: I see. It's kind of tricky to get `DispatchInit` to work with the launch helper so I removed this.

File: cub/cub/grid/grid_even_share.cuh [hunk 4]
```diff
@@ -129,10 +129,19 @@ public:
...
+      this->block_offset = 0;
+      this->block_end    = 0;
+      return;
+    }
+
     this->block_offset      = num_items_; // Initialize past-the-end
     this->block_end         = num_items_; // Initialize past-the-end
     this->num_items         = num_items_;
-    this->total_tiles       = _CUDA_VSTD::max(1, static_cast<int>(::cuda::ceil_div(num_items_, tile_items)));
+    this->total_tiles       = static_cast<int>(::cuda::ceil_div(num_items_, tile_items));
```
Comment (fbusato) [inline cub/cub/grid/grid_even_share.cuh]: this should be bounded by `INT_MAX`

File: cub/cub/grid/grid_even_share.cuh [hunk 5]
```diff
@@ -129,10 +129,20 @@ public:
...
+      this->block_offset = 0;
+      this->block_end    = 0;
+      return;
+    }
+
+    this->block_offset = num_items_; // Initialize past-the-end
+    this->block_end    = num_items_; // Initialize past-the-end
+    this->num_items    = num_items_;
+    this->total_tiles =
+      static_cast<int>(_CUDA_VSTD::min(static_cast<OffsetT>(INT_MAX), ::cuda::ceil_div(num_items_, tile_items)));
```
Comment (fbusato) [inline cub/cub/grid/grid_even_share.cuh]: sorry, I meant `::cuda::std::numeric_limits<int>::max()` + `cuda/std/limits` if missing

File: cub/cub/device/dispatch/dispatch_reduce_nondeterministic.cuh [hunk 6]
```diff
@@ -190,7 +190,7 @@ struct DispatchReduceNondeterministic
     GridEvenShare<OffsetT> even_share;
     even_share.DispatchInit(num_items, max_blocks, reduce_config.tile_size);
     // Get grid size for nondeterministic_device_reduce_atomic_kernel
-    const int reduce_grid_size = even_share.grid_size;
+    const int reduce_grid_size = _CUDA_VSTD::max(1, even_share.grid_size);
```
Comment (fbusato) [inline cub/cub/device/dispatch/dispatch_reduce_nondeterministic.cuh]:
```suggestion
    const int reduce_grid_size = ::cuda::std::max(1, even_share.grid_size);
```

File: cub/cub/grid/grid_even_share.cuh (hunk 5 above)
Comment (fbusato) [inline cub/cub/grid/grid_even_share.cuh]:
```suggestion
      static_cast<int>(::cuda::std::min(static_cast<OffsetT>(INT_MAX), ::cuda::ceil_div(num_items_, tile_items)));
```

**Review by fbusato — CHANGES_REQUESTED** [review]

File: cub/cub/grid/grid_even_share.cuh [hunk 7]
```diff
@@ -129,11 +130,21 @@ public:
...
+      return;
+    }
+
     this->block_offset      = num_items_; // Initialize past-the-end
     this->block_end         = num_items_; // Initialize past-the-end
     this->num_items         = num_items_;
-    this->total_tiles       = _CUDA_VSTD::max(1, static_cast<int>(::cuda::ceil_div(num_items_, tile_items)));
-    this->grid_size         = _CUDA_VSTD::min(total_tiles, max_grid_size);
+    this->total_tiles       = static_cast<int>(::cuda::std::min(
+      static_cast<OffsetT>(::cuda::std::numeric_limits<int>::max()), ::cuda::ceil_div(num_items_, tile_items)));
```
Comment (fbusato) [inline cub/cub/grid/grid_even_share.cuh]: nit:
```suggestion
      OffsetT{::cuda::std::numeric_limits<int>::max()}, ::cuda::ceil_div(num_items_, tile_items)));
```
